modules/tinymce/src/themes/silver/main/ts/ui/core/complex/BespokeTypeahead.ts

```diff
@@ -0,0 +1,181 @@
+          spec.onAction(item as any)(api);
+        }
+      };
+    })
+    : Arr.foldl(spec.dataset.getData(), (acc, item): Menu.ToggleMenuItemSpec[] => {
+      if (item.type === 'separator') {
+        return acc;
+      }
+      if (item.type === 'submenu') {
+        return acc.concat(Arr.map(item.items, (subItem) => ({ type: 'togglemenuitem', text: subItem.title, onAction: (api) => {
+          spec.onAction({ ...subItem, type: 'formatter' } as any)(api);
+        } })));
+      }
+
+      if (item.type === 'formatter') {
+        return acc.concat({ type: 'togglemenuitem', text: item.title, onAction: (api) => {
+          if (item.type === 'formatter') {
+            spec.onAction(item)(api);
+          }
+        } });
+      }
+      return acc;
+    },
+    [] as Menu.ToggleMenuItemSpec[]);
+
+  const onSetup = onSetupEvent(editor, 'NodeChange', (api: BespokeSelectApi) => {
+    const comp = api.getComponent();
+    spec.updateText(comp);
+  });
+  const getApi = (comp: AlloyComponent): BespokeSelectApi => ({ getComponent: Fun.constant(comp) });
+  const editorOffCell = Cell(Fun.noop);
+
+  const customEvents = Id.generate('custom-events');
+
+  const typeahead = Typeahead.sketch({
+    minChars: 1,
+    responseTime: 0.1,
+    markers: {
+      openClass: 'typeahead-is-open',
+    },
+    parts: {
+      menu: MenuParts.part(false, 1, 'normal')
+    },
+    model: {
+      selectsOver: true,
+      getDisplayText: (itemData: any) => itemData.meta && itemData.meta.text ? itemData.meta.text : 'No.text',
+      populateFromBrowse: true,
+    },
+    typeaheadBehaviours: Behaviour.derive([
+      AddEventsBehaviour.config(customEvents, [
+        onControlAttached({ onSetup, getApi }, editorOffCell),
+        onControlDetached({ getApi }, editorOffCell)
+      ]),
+      AddEventsBehaviour.config('menubutton-update-display-text', [
+        AlloyEvents.run<UpdateMenuTextEvent>(updateMenuText, (comp, se) => {
+          Representing.setValue(comp, {
+            meta: {
+              text: se.event.text,
+              format: se.event.format
+            }
+          });
+        })
+      ])
+    ]),
+    fetch: (_typeaheadComp) => {
+      const optTieredData = NestedMenus.build(
+        data,
+        ItemResponse.CLOSE_ON_EXECUTE,
+        backstage,
+        {
+          isHorizontalMenu: false,
+          search: Optional.none()
+        }
+      );
+      return Future.pure(optTieredData);
+    },
+    onExecute: (sandbox, item, value) => {
+      if (spec.dataset.type === 'advanced') {
+
+        const styles = spec.dataset.getData();
+
+        const selectedStyle = Arr.find(styles, (style) => {
+          if (style.type === 'formatter') {
+            return style.title === value.meta.text;
+          }
+          if (style.type === 'separator') {
+            return false;
+          }
+          if (style.type === 'submenu') {
+            return Arr.exists(style.items, (subItem) => subItem.title === value.meta.text);
```

Comment:
Are these styles going to be nested? Will you have to look inside subItem.items for example?